src/utils/EventUtils.cpp

```diff
@@ -0,0 +1,565 @@
+};
+
+
+// -----------------------------------------------------------------------
+
+EventData::EventData(std::string _name) :
+  name(_name)
+{}
+
+EventData::EventData(std::string _name, long _count, long _total,
+                     long _max, long _min, std::vector<int> _data, Event::StateChanges _stateChanges)
+  :  max(std::chrono::milliseconds(_max)),
+     min(std::chrono::milliseconds(_min)),
+     total(std::chrono::milliseconds(_total)),
+     stateChanges(_stateChanges),
+     name(_name),
+     count(_count),
+     data(_data)
+{}
+
+
+void EventData::put(Event* event)
+{
+  count++;
+  stdy_clk::duration duration = event->getDuration();
+  total += duration;
+  min = std::min(duration, min);
+  max = std::max(duration, max);
+  data.insert(std::end(data), std::begin(event->data), std::end(event->data));
+  stateChanges.insert(std::end(stateChanges), std::begin(event->stateChanges), std::end(event->stateChanges));
+}
+
+std::string EventData::getName() const
+{
+  return name;
+}
+
+long EventData::getAvg() const
+{
+  return (std::chrono::duration_cast<std::chrono::milliseconds>(total) / count).count();
+}
+
+long EventData::getMax() const
+{
+  return std::chrono::duration_cast<std::chrono::milliseconds>(max).count();
+}
+
+long EventData::getMin() const
+{
+  return std::chrono::duration_cast<std::chrono::milliseconds>(min).count();
+}
+
+long EventData::getTotal() const
+{
+  return std::chrono::duration_cast<std::chrono::milliseconds>(total).count();
+}
+
+long EventData::getCount() const
+{
+  return count;
+}
+
+std::vector<int> const & EventData::getData() const
+{
+  return data;
+}
+
+
+// -----------------------------------------------------------------------
+
+void RankData::initialize()
+{
+  initializedAt = sys_clk::now();
+  initializedAtTicks = stdy_clk::now();
+}
+
+void RankData::finalize()
+{
+  finalizedAt = sys_clk::now();
+  finalizedAtTicks = stdy_clk::now();
+}
+
+
+void RankData::put(Event* event)
+{
+  /// Construct or return EventData object with name as key and name as arg to ctor.
+  auto data = std::get<0>(evData.emplace(std::piecewise_construct,
+                                         std::forward_as_tuple(event->name),
+                                         std::forward_as_tuple(event->name)));
+  data->second.put(event);
+}
+
+
+void RankData::addEventData(EventData ed)
+{
+  evData.emplace(ed.getName(), std::move(ed));
+}
+
+
+void RankData::normalizeTo(sys_clk::time_point t0)
+{
+  auto const delta = initializedAt - t0; // duration that this rank initialized after the first rank
+  assertion(t0 <= initializedAt); // t0 should always be before or equal my init time
+
+  for (auto & events : evData) {
+    for (auto & sc : events.second.stateChanges) {
+      auto & tp = sc.second;
+      tp = stdy_clk::time_point(tp - initializedAtTicks + delta);
+      // Deactivated, as it fails for the precice tests, likely because Events are used before initialize
```

Comment:
I think this should be fine. If an Event was record before t0, then a negative tp is ok as our normalisation is more of an _alignement_.
```suggestion
```

Reply:
In an ideal world, there is no event recorded prior to t0, i.e., prior to initialization. But obviously, we do not live in an ideal world.